# ¡Hola, Cece!  

Mi nombre es Carlos Ortiz, soy code reviewer de Practicum y voy a revisar el proyecto que acabas de desarrollar.

Cuando vea un error la primera vez, lo señalaré. Deberás encontrarlo y arreglarlo. La intención es que te prepares para un espacio real de trabajo. En un trabajo, el líder de tu equipo hará lo mismo. Si no puedes solucionar el error, te daré más información en la próxima ocasión. 

Encontrarás mis comentarios más abajo - **por favor, no los muevas, no los modifiques ni los borres**.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Puedes responderme de esta forma: 

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div>

¡Empecemos!

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Propósito-del-Proyecto:" data-toc-modified-id="Propósito-del-Proyecto:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Propósito del Proyecto:</a></span></li><li><span><a href="#Hipótesis-a-evaluar:" data-toc-modified-id="Hipótesis-a-evaluar:-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Hipótesis a evaluar:</a></span></li><li><span><a href="#Carga-del-archivo:" data-toc-modified-id="Carga-del-archivo:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Carga del archivo:</a></span></li><li><span><a href="#Ejercicio-1.-Exploración-de-datos" data-toc-modified-id="Ejercicio-1.-Exploración-de-datos-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Ejercicio 1. Exploración de datos</a></span><ul class="toc-item"><li><span><a href="#Análisis-de-la-exploración-de-datos." data-toc-modified-id="Análisis-de-la-exploración-de-datos.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Análisis de la exploración de datos.</a></span></li></ul></li><li><span><a href="#Transformación-de-datos" data-toc-modified-id="Transformación-de-datos-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Transformación de datos</a></span></li><li><span><a href="#Trabajar-con-valores-ausentes" data-toc-modified-id="Trabajar-con-valores-ausentes-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Trabajar con valores ausentes</a></span><ul class="toc-item"><li><span><a href="#Restaurar-valores-ausentes-en-total_income" data-toc-modified-id="Restaurar-valores-ausentes-en-total_income-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Restaurar valores ausentes en <code>total_income</code></a></span></li><li><span><a href="#Restaurar-valores-en-days_employed" data-toc-modified-id="Restaurar-valores-en-days_employed-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Restaurar valores en <code>days_employed</code></a></span></li></ul></li><li><span><a href="#Clasificación-de-datos" data-toc-modified-id="Clasificación-de-datos-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Clasificación de datos</a></span></li><li><span><a href="#Comprobación-de-las-hipótesis" data-toc-modified-id="Comprobación-de-las-hipótesis-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Comprobación de las hipótesis</a></span></li><li><span><a href="#Conclusión-general" data-toc-modified-id="Conclusión-general-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Conclusión general</a></span></li></ul></div>

## Propósito del Proyecto:

El propósito de este proyecto es analizar sí, el estado civil y el número de hijos de los clientes tienen un impacto en el incumplimiento de pago de un préstamo. El banco cuenta con datos sobre la solvencia crediticia de los clientes, y el objetivo es determinar si estas variables pueden utilizarse para predecir la capacidad de un prestatario para pagar su préstamo.


## Hipótesis a evaluar:
Para este análisis, se formularán las siguientes hipótesis:

**Hipótesis 1:** Los clientes casados tienen una menor probabilidad de incumplir con los pagos de préstamos en comparación con los clientes solteros, divorciados o viudos.

**Hipótesis 2:** Los clientes con más hijos tienen una mayor probabilidad de incumplir con los pagos de préstamos en comparación con aquellos con menos o sin hijos.`

## Carga del archivo:



Se cargará el conjunto de datos y se realizará una exploración inicial para identificar valores faltantes y datos atípicos. Se describirá la estructura general de los datos.

In [1]:
import pandas as pd                     # Cargar todas las librerías
from IPython.display import display
import numpy as np



In [2]:
credit = pd.read_csv('/datasets/credit_scoring_eng.csv')

## Ejercicio 1. Exploración de datos

**Descripción de los datos**
- `children` - el número de hijos en la familia
- `days_employed` - experiencia laboral en días
- `dob_years` - la edad del cliente en años
- `education` - la educación del cliente
- `education_id` - identificador de educación
- `family_status` - estado civil
- `family_status_id` - identificador de estado civil
- `gender` - género del cliente
- `income_type` - tipo de empleo
- `debt` - ¿había alguna deuda en el pago de un préstamo?
- `total_income` - ingreso mensual
- `purpose` - el propósito de obtener un préstamo

Verifiquemos cuantas columnas y filas hay en el conjunto de datos.


In [3]:
credit.shape



(21525, 12)

In [4]:
 # vamos a mostrar las primeras filas N

credit.head()


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

### Análisis de la exploración de datos.

Al observar la muestra de datos mostrada en la pantalla, noto lo siguiente:

 1. Las columnas "children", "dob_years", "education", "family_status",  "gender", debt e "income_type" parecen contener valores válidos, corresponde con su tipo de datos.

 2. La columna "days_employed" muestra la experiencia laboral del cliente en días. Sin embargo, noto que algunos valores son negativos y otros son extremadamente altos, como el valor 340266.072047 en la fila 4. Esto parece indicar que hay problemas con los datos en esta columna que requerirían investigación y posiblemente cambios adicionales.

 3. La columna "total_income" representa el ingreso mensual del cliente. Al igual que con la columna "days_employed", noto que hay valores ausentes en esta columna que requerirían investigación y posiblemente cambios adicionales.

En resumen, hay algunos problemas en las columnas "days_employed" y "total_income" que necesitan ser investigados y tratados, ya que contienen valores negativos y ausentes.  Además, sería importante verificar si hay valores atípicos o errores en otras columnas y abordarlos si es necesario para garantizar la calidad y precisión de los datos antes de realizar análisis adicionales.


In [5]:
# Obtener información sobre los datos
display(credit.info())
display(credit.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


None

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

 
 
Obtuve información de los datos usando el método info en donde observo una tabla que contiene 12 Columnas con un total de 21525 registros.

Se observa las siguientes problemas:
1. Corroboramos que las columnas  'days_employed' y 'total_income' contienen valores ausentes.
2. En la descripción de los datos se observa valores negativos en children y days_employed.
3. valores atipicos como el máximo de días trabajados en la variable de days_employed.

In [6]:
# Veamos la tabla filtrada con valores ausentes de la primera columna donde faltan datos
credit_day_na = credit[credit['days_employed'].isna()]
display(credit_day_na)


children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [7]:
credit_na = credit[credit['days_employed'].isna()&
             credit['total_income'].isna()]
credit_na.shape[0]                

2174

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Bien hecho.
</div>

La columna "days_employed" tiene valores ausentes (NaN) para varias filas de la tabla.

También parece que la columna "total_income" tiene valores ausentes (NaN) para las mismas filas que la columna "days_employed".

Otras columnas también tienen valores ausentes (NaN) en algunas filas, como "total_income", "purpose", entre otras.

Es interesante notar que las filas con valores ausentes en "days_employed" también tienen valores ausentes en "total_income". Esto puede indicar que hay un problema con la recolección de datos o alguna otra razón por la cual estos valores están faltando en conjunto.

Las filas con valores ausentes en "days_employed" y "total_income" también tienen valores ausentes en otras columnas, como "gender", "family_status_id", "debt", entre otras.

Dado que varios campos tienen valores ausentes en estas filas, es importante investigar la causa de estos valores faltantes y determinar cómo manejarlos adecuadamente para el análisis de datos. Algunas posibles acciones a considerar son la imputación de valores ausentes utilizando la mediana o la media, la eliminación de las filas con valores ausentes o el uso de técnicas de análisis más avanzadas para tratar con datos faltantes.

In [8]:
filtered_data = credit[(credit['days_employed'].isna()) & (credit['total_income'].isna())]
display(filtered_data.shape[0])

2174

In [9]:
day_employed_nan = credit['days_employed'].isna().sum()

# Calcula el porcentaje de valores ausentes
porcentaje_nan = (day_employed_nan / len(credit)) * 100
porcentaje_nan

10.099883855981417

**Conclusión intermedia**

El número de filas en la tabla filtrada (2174) coincide con el número de valores ausentes en las columnas "days_employed" y "total_income". Esto indica que todos los registros que tienen datos faltantes en una de estas columnas también tienen datos faltantes en la otra columna.

Este porcentaje (aproximadamente 10.09%) indica que una porción significativa de los datos tiene valores ausentes. Aunque no es un valor extremo, justifica la necesidad de abordarlos razonablmente.
Hay que examinar las caracteristicas del cliente para poder entender si la razón de estos faltantes es que siguen algún patrón o son aleatorios.

Los próximos pasos implicarían:

 1. Identificar posibles razones para los valores ausentes en las columnas "days_employed" y "total_income" mediante el análisis de características específicas de los clientes.
 2. Verificar si la presencia de valores ausentes en estas columnas está correlacionada con otras características de los clientes.
3. Decidir una estrategia de imputación para completar los valores ausentes. Esto podría implicar utilizar la mediana o la media de la columna.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Muy buen trabajo realizado en esta sección. 
</div>

In [10]:
# Vamos a investigar a los clientes que no tienen datos sobre la característica identificada y la columna con los valores ausentes

clientes_nan = credit[credit['days_employed'].isna()]

# Imprimir el número de filas en la tabla filtrada
display("Número de filas con valores ausentes en 'days_employed':", len(clientes_nan))
clientes_nan.head()

"Número de filas con valores ausentes en 'days_employed':"

2174

children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   

    total_income                   purpose  
12           NaN         to have a wedding  
26           NaN                 education  
29           NaN    building a real estate  
41           NaN  second-hand car purchase  
55           NaN         to have a wedding

In [11]:
# Miremos cómo se distribuyen los valores de 'income_type' en las filas con valores faltantes en 'days_employed'.
credit_day_employed_nan = credit.loc[credit['days_employed'].isna(),'income_type'].value_counts(normalize= True)
credit_day_employed_nan

employee         0.508280
business         0.233671
retiree          0.189972
civil servant    0.067617
entrepreneur     0.000460
Name: income_type, dtype: float64

Esto significa que el 51% de las filas donde falta el valor en 'days_employed' tienen 'income_type' como 'employee', el 23% tienen 'income_type' como 'business'.

Este análisis muestra como se distribuyen los diferentes tipos de ingresos en las filas donde falta la información de 'days_employed', lo que podría proporcionar pistas sobre la naturaleza de los valores ausentes y su posible relación con el tipo de ingreso del cliente.

**Conclusión intermedia**


Los valores ausentes en la columna 'total_income' parecen estar relacionados con la falta de información sobre la experiencia laboral ('days_employed'), ya que el número de filas con valores faltantes en ambas columnas coincide. Esto sugiere que los datos ausentes pueden estar relacionados con la falta de ingresos declarados debido a la falta de información laboral.

La distribución de los diferentes tipos de ingresos ('income_type') en las filas con valores faltantes en 'days_employed' muestra que el grupo más representativo es el de empleados ('employee') con alrededor del 51% de las filas con valores faltantes. Esto podría indicar que los valores ausentes en 'days_employed' pueden estar más asociados con los empleados que con otros tipos de ingresos.

Según el análisis realizado hasta ahora, podemos identificar algunas posibles razones por las cuales hay valores ausentes:

 1. Ingresos no declarados: Los valores ausentes en la columna 'total_income' podrían deberse a que algunos clientes no declararon su ingreso total al momento de proporcionar la información.

2. Desempleo: Los valores ausentes en la columna 'days_employed' podrían indicar que algunos clientes están desempleados y, por lo tanto, no tienen información sobre su experiencia laboral.

3. Errores de registro: Es posible que algunos valores ausentes sean el resultado de errores de registro.


Para comprobar si los valores ausentes son aleatorios o si hay algún patrón, podríamos realizar análisis adicionales para investigar la relación entre la falta de información laboral ('days_employed') y otras caracteristicas, como el estado civil, el tipo de empleo, la educación.
 

In [12]:
income_type_distribution = credit['income_type'].value_counts(normalize=True)
income_type_distribution

employee                       0.516562
business                       0.236237
retiree                        0.179141
civil servant                  0.067782
unemployed                     0.000093
entrepreneur                   0.000093
student                        0.000046
paternity / maternity leave    0.000046
Name: income_type, dtype: float64

**Conclusión intermedia**

La distribución de los tipos de empleo en el conjunto de datos original es similar a la distribución en la tabla filtrada de valores ausentes en 'days_employed', lo que significa que la presencia de valores ausentes en 'days_employed' no parece depender significativamente del tipo de empleo. Esto sugiere que los valores ausentes en 'days_employed' pueden no estar relacionados directamente con el tipo de empleo del cliente.

podríamos comparar los valores de 'days_employed' entre diferentes grupos de clientes para ver si hay alguna diferencia significativa en la experiencia laboral en función de otras características. Por ejemplo, podríamos comparar los días de empleo promedio entre los clientes con diferentes niveles de educación o entre los clientes con diferentes propósitos de préstamo.

In [13]:
# Agrupar por 'education' y 'purpose' y calcular el promedio de 'days_employed' para cada grupo
days_employed_by_education_and_purpose = credit.groupby(['education', 'purpose'])['days_employed'].mean()
display("Días de empleo promedio por nivel de educación y propósito de préstamo:")
display(days_employed_by_education_and_purpose)

'Días de empleo promedio por nivel de educación y propósito de préstamo:'

education          purpose                                 
BACHELOR'S DEGREE  building a property                          -5156.581515
                   building a real estate                       67510.208031
                   buy commercial real estate                   -4382.444558
                   buy real estate                             136497.539116
                   buy residential real estate                  -1335.591132
                                                                   ...      
some college       to own a car                                 -1349.718176
                   transactions with commercial real estate     -1167.545171
                   transactions with my real estate             15065.596829
                   university education                         -2468.220788
                   wedding ceremony                             -1827.402290
Name: days_employed, Length: 386, dtype: float64

**Conclusión intermedia**

No hay datos suficientes para calcular el promedio: Puede que para ciertas combinaciones de nivel de educación y propósito de préstamo no haya suficientes datos disponibles en la columna 'days_employed' para calcular un promedio significativo.

Los valores ausentes están presentes en todas las filas: Si todos los valores de 'days_employed' para una combinación específica de nivel de educación y propósito de préstamo son valores ausentes (NaN), no se puede calcular un promedio válido.

In [14]:
# Funcion para comparar la distribución de valores faltantes en la columnas con valores ausentes con la distribución del dataframe completa

columns_to_verify = [c for c in credit.columns
                    if c not in ('total_income','days_employed')]

def check_distribution(dataframe,column):
    full_counts = dataframe[column].value_counts(normalize= True)
    not_na_counts = dataframe.loc[dataframe['total_income'].isna(), column].value_counts(normalize = True)
    absdelta = (full_counts - not_na_counts).abs()
    
    return pd.DataFrame([full_counts, not_na_counts,absdelta], index=['full','not_na','abs_delta'])

In [15]:
# Verificando la funcion.
check_distribution (credit,'children')

-1         0         1         2         3         4   \
full       0.002184  0.657329  0.223833  0.095470  0.015331  0.001905   
not_na     0.001380  0.661914  0.218491  0.093836  0.016559  0.003220   
abs_delta  0.000804  0.004585  0.005341  0.001634  0.001228  0.001315   

                 5         20  
full       0.000418  0.003531  
not_na     0.000460  0.004140  
abs_delta  0.000042  0.000609

Identifiquemos qué columnas tienen la mayor discrepancia en la distribución de valores entre las filas con valores ausentes en 'total_income' y el DataFrame completo. 

In [16]:
data = []

for column in columns_to_verify:
    results = check_distribution(credit,column)
    deltas = results.loc['abs_delta']
    biggest_change_name = deltas.idxmax()
    biggest_change_value = deltas[biggest_change_name]
    text = f" Máximo cambio porcentual:{biggest_change_value:.02%} (en {column}=='{biggest_change_name}')"
    data.append((column,biggest_change_name,biggest_change_value,text)) 
 
changes = pd.DataFrame(data, columns =['column','biggest_change_in','biggest_change_value','description'])
changes                                                                     
                                                                     

column  biggest_change_in  biggest_change_value  \
0          children                  1              0.005341   
1         dob_years                 32              0.006674   
2         education  bachelor's degree              0.008964   
3      education_id                  0              0.005863   
4     family_status  civil partnership              0.009258   
5  family_status_id                  1              0.009258   
6            gender                  F              0.021242   
7       income_type            retiree              0.010832   
8              debt                  0              0.002686   
9           purpose       to buy a car              0.008129   

                                         description  
0   Máximo cambio porcentual:0.53% (en children==...  
1   Máximo cambio porcentual:0.67% (en dob_years=...  
2   Máximo cambio porcentual:0.90% (en education=...  
3   Máximo cambio porcentual:0.59% (en education_...  
4   Máximo cambio porcentual:0.93% (en family_sta...  
5   Máximo cambio porcentual:0.93% (en family_sta...  
6    Máximo cambio porcentual:2.12% (en gender=='F')  
7   Máximo cambio porcentual:1.08% (en income_typ...  
8      Máximo cambio porcentual:0.27% (en debt=='0')  
9   Máximo cambio porcentual:0.81% (en purpose=='...

**Conclusiones**

**Patrones encontrados:** Se observaron diferencias significativas en la distribución de valores entre las filas con valores ausentes en la columna 'total_income' y el DataFrame completo en diversas características de los clientes, como el número de hijos, la edad, el nivel educativo, el estado civil, el género, el tipo de empleo, la presencia de deudas y el propósito del préstamo. Estos patrones sugieren que los valores ausentes pueden no ser aleatorios y podrían estar relacionados con características específicas de los clientes.

**Abordar los valores ausentes:** Dado que los valores ausentes pueden estar relacionados con características específicas de los clientes, una estrategia para manejarlos sería completar los valores ausentes utilizando la mediana o la media de 'total_income' y 'days_employed'

Próximos pasos en la transformación de datos: Manejar duplicados, completar valores ausentes y verificación de los datos.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Excelente trabajo en esta sección. Abordaste las columnas e identificaste posibles patrones de los datos ausentes. En un futuro podrás investigar más a fondo si tus datos son MAR (missing at random), MCAR (missing completely at random) o MNAR (missing not at random).
</div>

## Transformación de datos


In [17]:
# Veamos todos los valores en la columna de educación para verificar si será necesario corregir la ortografía y qué habrá que corregir exactamente
credit['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [18]:
# Arregla los registros si es necesario

credit['education'] = credit['education'].str.lower()
credit['education'].value_counts()


secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

In [19]:
credit['education'].replace({
    "secondary education": "secondary_education",
    "some college": "some_college",
    "bachelor's degree": "bachelor_degree",
    "primary education": "primary_education",
    "graduate degree": "graduate_degree"
}, inplace=True)

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo.
</div>

In [20]:
# Comprobar todos los valores en la columna para asegurarnos de que los hayamos corregido

credit['education'].value_counts()

secondary_education    15233
bachelor_degree         5260
some_college             744
primary_education        282
graduate_degree            6
Name: education, dtype: int64

**Verificando datos de la columna 'children'**

In [21]:
# Veamos la distribución de los valores en la columna `children`
children_distribution = credit['children'].value_counts(normalize= True)
children_distribution

 0     0.657329
 1     0.223833
 2     0.095470
 3     0.015331
 20    0.003531
-1     0.002184
 4     0.001905
 5     0.000418
Name: children, dtype: float64

En la columna children, se observan dos valores que pueden considerarse extraños y potencialmente problemáticos: el valor 20 y el valor -1. Ambos valores son atípicos y poco probables en el contexto del número de hijos en una familia.

El valor 20 representa aproximadamente el 0.35% de los datos, mientras que el valor -1 representa aproximadamente el 0.22% de los datos. En conjunto, estos dos valores problemáticos representan aproximadamente el 0.57% de los datos en la columna children.

Es posible que estos datos incorrectos hayan ocurrido debido a errores humanos o problemas en el proceso de recopilación de datos. 

Dado que estos valores representan solo un pequeño porcentaje del conjunto de datos, una opción razonable sería o corregirlos. Podemos reemplazar el valor 20 con 2 (asumiendo que es un error tipográfico) y modificar el valor -1 asignándole la mediana de toda la columna.

In [22]:
# Reemplazar el valor 20 con 2
credit['children'] = credit['children'].replace(20, 2)

# Obtener la mediana de la columna 'children' excluyendo el valor -1
children_median = credit[credit['children'] != -1]['children'].median()

# Reemplazar los valores -1 en la columna 'children' con la mediana
credit['children'] = credit['children'].replace(-1, children_median)                                                         

In [23]:
# Comprobar la columna `children` de nuevo para asegurarnos de que todo está arreglado
children_distribution = credit['children'].value_counts(normalize=True)
children_distribution

0.0    0.659512
1.0    0.223833
2.0    0.099001
3.0    0.015331
4.0    0.001905
5.0    0.000418
Name: children, dtype: float64

**Verifiquemos la columna 'days_Employed'**

En el análisis realizado a la columna days_employed, podemos identificar algunos problemas potenciales:

Valores negativos: Los valores negativos en days_employed no tienen sentido, ya que la experiencia laboral no puede ser negativa. Es posible que estos valores sean errores o datos incorrectos.

Valores extremadamente grandes: También podemos encontrar valores extremadamente grandes en days_employed, lo que podría indicar datos incorrectos en el conjunto de datos.

In [24]:
# Encuentra datos problemáticos en `days_employed`, si existen, y calcula el porcentaje
days_employed_check = credit['days_employed'].describe()
days_employed_check

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

In [25]:
# Contar la cantidad de datos problemáticos tomando como referencia un rango valores negativos y un limite de 50 años (18250 dias)
days_employed_atipicos = len(credit[(credit['days_employed'] < 0) | (credit['days_employed'] > 18250)])

# Calcular el porcentaje de datos problemáticos
total_credit = len(credit['days_employed'])
porcentage_atipico = (days_employed_atipicos / total_credit) * 100

print("Porcentaje de datos problemáticos en 'days_employed': {:.2f}%".format(porcentage_atipico))

Porcentaje de datos problemáticos en 'days_employed': 89.90%


El 89.89% es un valor alto, es posible que se deba a un error en el proceso de registro de datos o en la codificación de los valores.

Para no eliminar las filas problemáticas, podríamos considerar dos enfoques para la imputación de los valores problemáticos en la columna 'days_employed':

Reemplazar los valores negativos con sus valores absolutos: Si los valores negativos no tienen sentido en el contexto de la columna 'days_employed', podemos simplemente tomar el valor absoluto de esos valores para corregirlos.

Establecer un límite razonable para los días empleados: Dado que los valores extremadamente altos son inverosímiles, podemos establecer un límite superior razonable, como 50 años (aproximadamente 18250 días), y reemplazar los valores que excedan ese límite con el límite establecido.


In [26]:

# Reemplazando los valores negativos por valor absoluto
credit['days_employed'] = credit['days_employed'].apply(lambda x: abs(x) if x < 0 else x)


In [27]:
# Estableciendo un límite razonable para los días empleados (por ejemplo, 50 años):
upper_limit = 50 * 365  # 50 años en días
credit['days_employed'] = credit['days_employed'].apply(lambda x: upper_limit if x > upper_limit else x)

In [28]:
credit['days_employed'].describe()

count    19351.000000
mean      5183.100742
std       6430.122757
min         24.141633
25%        927.009265
50%       2194.220567
75%       5537.882441
max      18250.000000
Name: days_employed, dtype: float64

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Muy buen análisis.
</div>

**Verificando la edad de los clientes 'dob_years'**

Observemos la distribución de las edades de los clientes en el conjunto de datos. Veamos la frecuencia de cada edad en orden ascendente:

In [29]:
# Revisa `dob_years` en busca de valores sospechosos y cuenta el porcentaje
dob_years_counts = credit['dob_years'].value_counts(normalize = True).sort_index(ascending=True)
dob_years_counts


0     0.004692
19    0.000650
20    0.002369
21    0.005157
22    0.008502
23    0.011800
24    0.012265
25    0.016585
26    0.018955
27    0.022904
28    0.023368
29    0.025319
30    0.025087
31    0.026016
32    0.023693
33    0.026992
34    0.028014
35    0.028664
36    0.025784
37    0.024948
38    0.027782
39    0.026620
40    0.028293
41    0.028200
42    0.027735
43    0.023833
44    0.025412
45    0.023089
46    0.022067
47    0.022300
48    0.024994
49    0.023600
50    0.023879
51    0.020813
52    0.022485
53    0.021324
54    0.022253
55    0.020581
56    0.022625
57    0.021370
58    0.021417
59    0.020627
60    0.017515
61    0.016492
62    0.016353
63    0.012497
64    0.012311
65    0.009013
66    0.008502
67    0.007758
68    0.004599
69    0.003949
70    0.003020
71    0.002695
72    0.001533
73    0.000372
74    0.000279
75    0.000046
Name: dob_years, dtype: float64

La columna dob_years parece tener una distribución razonable de edades, con valores que aumentan progresivamente a medida que los clientes son más jóvenes y disminuyen hacia edades más avanzadas. Sin embargo, hay algunos valores sospechosos:

Hay valores con edad igual a 0, lo cual es poco probable ya que generalmente se requiere que los clientes tengan al menos 18 años para acceder a productos financieros como préstamos.



In [30]:
# Vamos a calcular el porcentaje de valores sospechosos para las edades 0 y mayores o iguales a 85 años:

age_check = credit[(credit['dob_years'] == 0) | (credit['dob_years'] >= 85)]
percentage_age_check = len(age_check) / len(credit['dob_years']) * 100

print(f"Porcentaje de valores sospechosos en 'dob_years': {percentage_age_check:.2f}%")

Porcentaje de valores sospechosos en 'dob_years': 0.47%


Para mantener más datos y preservar la información útil de otras columnas, imputando valores en lugar de eliminar los registros tomando variables como  (education e income_type) podría corregir los datos atipicos  

In [31]:
# Calculando la mediana de edad por grupos de 'education' y 'income_type'
median_age_by_group = credit.groupby(['education', 'income_type'])['dob_years'].median()

# Función para imputar la edad basada en la mediana del grupo
def age_median (row):
    if row['dob_years'] == 0 or row['dob_years'] >= 85:
        return median_age_by_group.loc[(row['education'], row['income_type'])]
    return row['dob_years']


credit['dob_years'] = credit.apply(age_median, axis=1)

In [32]:

# Verificar los cambios realizados
credit['dob_years'] = credit['dob_years'].astype(int)
dob_years_counts = credit['dob_years'].value_counts(normalize = True).sort_index(ascending=True)
dob_years_counts

19    0.000650
20    0.002369
21    0.005157
22    0.008502
23    0.011800
24    0.012265
25    0.016585
26    0.018955
27    0.022904
28    0.023368
29    0.025319
30    0.025087
31    0.026109
32    0.023693
33    0.026992
34    0.028014
35    0.028664
36    0.025784
37    0.025784
38    0.028293
39    0.026620
40    0.030105
41    0.028618
42    0.027828
43    0.023833
44    0.025412
45    0.023089
46    0.022067
47    0.022300
48    0.024994
49    0.023600
50    0.023879
51    0.020813
52    0.022485
53    0.021324
54    0.022253
55    0.020581
56    0.022625
57    0.021370
58    0.021417
59    0.020906
60    0.018165
61    0.016492
62    0.016353
63    0.012497
64    0.012311
65    0.009013
66    0.008502
67    0.007758
68    0.004599
69    0.003949
70    0.003020
71    0.002695
72    0.001533
73    0.000372
74    0.000279
75    0.000046
Name: dob_years, dtype: float64

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Muy buen trabajo.
</div>

**Verificar la columna 'family_status'**

In [33]:
# Veamos los valores de la columna 'family_status'

credit['family_status'].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

Los valores obtenidos parecen ser validos, por lo que no se considera hacer ningún cambio.
Por el momento realizaré una actualización de los valores eliminando espacios.

In [34]:
# eliminando espacios en los valores
credit['family_status'].replace({
    "civil partnership": "civil_partnership",
    "widow / widower": "widow_widower"
}, inplace=True)


In [35]:
# Verificando el resultado
credit['family_status'].value_counts()

married              12380
civil_partnership     4177
unmarried             2813
divorced              1195
widow_widower          960
Name: family_status, dtype: int64

**Verificar la columna 'gender'**

In [36]:
# Veamos los valores en la columna 'gender'
credit['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

No se evidencia datos incoherentes, el valor XNA se puede interpretar como sexo no determinado y podría ser valido. además que un registro no alteraría el análisis.

**Verificar la columna 'income_type'**

In [37]:
# Veamos los valores en la columna 'income_type'
credit['income_type'].value_counts()

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

Los valores obtenidos parecen ser validos, por lo que no se considera hacer ningún cambio.
Por el momento realizaré una actualización de los valores eliminando espacios.

In [38]:

# eliminando espacios en los valores
credit['income_type'].replace({
    "civil servant": "civil_servant",
    "paternity / maternity leave": "parent_leave"
}, inplace=True)


In [39]:
# Veamos los valores en la columna 'income_type'
credit['income_type'].value_counts()

employee         11119
business          5085
retiree           3856
civil_servant     1459
unemployed           2
entrepreneur         2
parent_leave         1
student              1
Name: income_type, dtype: int64

**Verificar Duplicados**


In [40]:
# Comprobar los duplicados

duplicated_credit = credit[credit.duplicated()]
display(duplicated_credit.head())
display(credit.duplicated().sum())

children  days_employed  dob_years            education  education_id  \
2849       0.0            NaN         41  secondary_education             1   
3290       0.0            NaN         58  secondary_education             1   
4182       1.0            NaN         34      bachelor_degree             0   
4851       0.0            NaN         60  secondary_education             1   
5557       0.0            NaN         58  secondary_education             1   

          family_status  family_status_id gender income_type  debt  \
2849            married                 0      F    employee     0   
3290  civil_partnership                 1      F     retiree     0   
4182  civil_partnership                 1      F    employee     0   
4851  civil_partnership                 1      F     retiree     0   
5557  civil_partnership                 1      F     retiree     0   

      total_income                              purpose  
2849           NaN  purchase of the house for my family  
3290           NaN                    to have a wedding  
4182           NaN                     wedding ceremony  
4851           NaN                     wedding ceremony  
5557           NaN                    to have a wedding

71

In [41]:
# Eliminando duplicados.
credit = credit.drop_duplicates().reset_index(drop=True)

In [42]:
# Última comprobación para ver si tenemos duplicados
display(credit.duplicated().sum())

0

In [43]:
# Comprueba el tamaño del conjunto de datos que tienes ahora, después de haber ejecutado estas primeras manipulaciones
credit.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  float64
 1   days_employed     19351 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21454 non-null  object 
dtypes: float64(3), int64(4), object(5)
memory usage: 2.0+ MB


Después de realizar las correcciones y la imputación de valores, el nuevo conjunto de datos tiene 21,407 filas y 12 columnas. Los cambios realizados incluyen:

1. Imputación de valores faltantes en la columna 'days_employed'
2. Imputación de valores faltantes en la columna 'dob_years'
3. Eliminacion de duplicados.

En general, se han realizado ajustes en aproximadamente el 20.10% del conjunto de datos original para abordar los valores faltantes y sospechosos. Estos cambios deberían ayudar a mejorar la calidad y confiabilidad del análisis de los datos.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Gran trabajo en esta sección.
</div>

## Trabajar con valores ausentes

Trabajar con diccionarios puede acelerar y facilitar el procesamiento de datos, especialmente cuando se tienen columnas con valores codificados o representados mediante IDs en lugar de texto legible. Al utilizar diccionarios, podemos mapear los IDs a sus respectivos valores descriptivos, lo que nos permite comprender mejor y analizar los datos de manera más eficiente.

En el conjunto de datos proporcionado, hay algunas columnas que contienen valores codificados con IDs, como 'education_id', 'family_status_id', y 'income_type'. En lugar de trabajar directamente con los IDs, podemos crear diccionarios que relacionen cada ID con su descripción correspondiente.

In [44]:
# Encuentra los diccionarios

family_status = {
    0: 'civil_partnership',
    1: 'married',
    2: 'widow_widower',
    3: 'divorced',
    4: 'unmarried',
}



education = {
    0: 'secondary_education',
    1: "bachelor_degree",
    2: 'some_college',
    3: 'primary_education',
    4: 'graduate_degree',
}

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Has identificado los diccionarios correctamente.
</div>

### Restaurar valores ausentes en `total_income`

Las columnas que tienen valores ausentes que debemos abordar son 'days_employed' y 'total_income'.

Para abordar los valores ausentes en 'days_employed', utilizaremos un enfoque de imputación. Reemplazaremos los valores negativos con sus valores absolutos y estableceremos un límite razonable para los días empleados (por ejemplo, 50 años) para reemplazar los valores extremadamente altos.

En cuanto a los valores ausentes en 'total_income', utilizaremos la mediana de ingresos para cada grupo de 'income_type' y 'education' para reemplazar los valores faltantes. Esto nos permitirá conservar la distribución de ingresos en cada grupo y evitar sesgos en los datos.

In [45]:
# Verificando la edad de los clientes
display (credit['dob_years'].value_counts().sort_index(ascending=True))


19     14
20     51
21    111
22    183
23    252
24    264
25    357
26    408
27    493
28    503
29    544
30    537
31    561
32    509
33    581
34    601
35    616
36    554
37    554
38    608
39    572
40    646
41    614
42    598
43    512
44    545
45    496
46    472
47    477
48    536
49    508
50    513
51    446
52    484
53    459
54    476
55    443
56    483
57    456
58    454
59    449
60    388
61    354
62    348
63    269
64    260
65    193
66    182
67    167
68     99
69     85
70     65
71     56
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [46]:
# Vamos a escribir una función que calcule la categoría de edad

def age_group(dob_years):
    if dob_years <= 17 or pd.isna(dob_years):
        return 'NA'
    elif dob_years < 20:
        return '18-19'
    elif dob_years < 30:
        return '20-29'
    elif dob_years < 40:
        return '30-39'
    elif dob_years < 50:
        return '40-49'
    elif dob_years < 60:
        return '50-59'
    elif dob_years < 70:
        return '60-69'
    else:
        return '70+'
    


In [47]:
# Prueba si la función funciona bien
age_group(0)

'NA'

In [48]:
# Crear una nueva columna basada en la función

credit['age_group'] = credit['dob_years'].apply(age_group) 

    

In [49]:
# Comprobar cómo los valores en la nueva columna

credit.head()

children  days_employed  dob_years            education  education_id  \
0       1.0    8437.673028         42      bachelor_degree             0   
1       1.0    4024.803754         36  secondary_education             1   
2       0.0    5623.422610         33  secondary_education             1   
3       3.0    4124.747207         32  secondary_education             1   
4       0.0   18250.000000         53  secondary_education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil_partnership                 1      F     retiree     0     25378.572   

                   purpose age_group  
0    purchase of the house     40-49  
1             car purchase     30-39  
2    purchase of the house     30-39  
3  supplementary education     30-39  
4        to have a wedding     50-59

En general, es importante tener en cuenta la naturaleza de los datos y las características de la distribución antes de tomar una decisión sobre si utilizar la media o la mediana para imputar los valores ausentes en 'total_income'

In [50]:
# Crea una tabla sin valores ausentes y muestra algunas de sus filas para asegurarte de que se ve bien
credit_no_nan = credit.dropna(inplace=False)
credit_no_nan.head()

children  days_employed  dob_years            education  education_id  \
0       1.0    8437.673028         42      bachelor_degree             0   
1       1.0    4024.803754         36  secondary_education             1   
2       0.0    5623.422610         33  secondary_education             1   
3       3.0    4124.747207         32  secondary_education             1   
4       0.0   18250.000000         53  secondary_education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil_partnership                 1      F     retiree     0     25378.572   

                   purpose age_group  
0    purchase of the house     40-49  
1             car purchase     30-39  
2    purchase of the house     30-39  
3  supplementary education     30-39  
4        to have a wedding     50-59

In [51]:
numeric_columns = ['total_income']

# Calcular las medias agrupadas por 'income_type' para las columnas en numeric_columns
mean_values = credit.groupby(['children','income_type'])[numeric_columns].mean()

# Calcular las medianas agrupadas por 'income_type' para las columnas en numeric_columns
median_values = credit.groupby(['children','income_type'])[numeric_columns].median()



# Comprobando el resultado
display('mean:')
display(mean_values)
display()
display('medians:')
display(median_values)

'mean:'

total_income
children income_type                
0.0      business       32385.039986
         civil_servant  27939.766263
         employee       25823.504156
         entrepreneur   79866.103000
         retiree        21792.006266
         student        15712.260000
         unemployed     32435.602000
1.0      business       32228.141977
         civil_servant  26151.320474
         employee       25777.135121
         retiree        23286.985710
         unemployed      9593.119000
2.0      business       32610.332073
         civil_servant  26365.038181
         employee       25591.072475
         parent_leave    8612.661000
         retiree        26391.040250
3.0      business       33816.952569
         civil_servant  29140.416917
         employee       27695.360922
         retiree        25301.815667
4.0      business       23488.481500
         civil_servant  31861.001000
         employee       26996.704630
         retiree        24522.216000
5.0      business       29816.225500
         civil_servant  20176.344000
         employee       27668.396600

'medians:'

total_income
children income_type                
0.0      business         27929.0980
         civil_servant    24969.7880
         employee         22872.1825
         entrepreneur     79866.1030
         retiree          18781.1670
         student          15712.2600
         unemployed       32435.6020
1.0      business         26909.7545
         civil_servant    22553.6195
         employee         22944.8310
         retiree          20051.6480
         unemployed        9593.1190
2.0      business         27387.2650
         civil_servant    22781.1910
         employee         22160.9305
         parent_leave      8612.6610
         retiree          22359.8005
3.0      business         27835.9060
         civil_servant    25909.3695
         employee         23113.1695
         retiree          18953.4285
4.0      business         23488.4815
         civil_servant    27209.0255
         employee         25210.9740
         retiree          24522.2160
5.0      business         29816.2255
         civil_servant    20176.3440
         employee         34007.2590

Imputar los valores ausentes en 'total_income' utilizando la mediana de los ingresos agrupados por 'children' e 'income_type' para obtener una estimación más representativa del ingreso de cada cliente en función de su situación particular de 'children' e 'income_type'.


In [52]:
# Crear una tabla externa con las medianas de 'total_income' para cada combinación de 'income_type' y 'children'
total_income_median = credit.groupby(['income_type', 'children'])['total_income'].median()

# Función para imputar los valores ausentes en 'total_income'
def total_income_nan(row):
    income_type = row['income_type']
    children = row['children']
    median_income = total_income_median.loc[income_type, children]
    
    if pd.isna(row['total_income']):
        return median_income
    else:
        return row['total_income']



In [53]:

credit['total_income'] = credit.apply(total_income_nan, axis=1)

In [54]:
# Comprueba si tenemos algún error
credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  float64
 1   days_employed     19351 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  age_group         21454 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 2.1+ MB


In [55]:
# Comprobar el número de entradas en las columnas

credit.count()

children            21454
days_employed       19351
dob_years           21454
education           21454
education_id        21454
family_status       21454
family_status_id    21454
gender              21454
income_type         21454
debt                21454
total_income        21454
purpose             21454
age_group           21454
dtype: int64

###  Restaurar valores en `days_employed`

Para restaurar los valores ausentes en la columna 'days_employed', podemos seguir un enfoque similar al que utilizamos para restaurar los valores en la columna 'total_income'. Es decir, agrupar los datos por características relevantes y calcular la mediana de 'days_employed' dentro de cada grupo.

In [56]:
numeric_columns = [ 'days_employed']

# Calcular las medias agrupadas por 'income_type' para las columnas en numeric_columns
mean_values = credit.groupby(['education_id','income_type'])[numeric_columns].mean()

# Calcular las medianas agrupadas por 'income_type' para las columnas en numeric_columns
median_values = credit.groupby(['education_id','income_type'])[numeric_columns].median()



# Comprobando el resultado
display('mean:')
display(mean_values)
display()
display('medians:')
display(median_values)


'mean:'

days_employed
education_id income_type                 
0            business         2018.687160
             civil_servant    3182.314561
             employee         2266.736782
             entrepreneur      520.848083
             retiree         18250.000000
             student           578.751554
             unemployed      18250.000000
1            business         2235.658250
             civil_servant    3604.210297
             employee         2384.626455
             parent_leave     3296.759962
             retiree         18250.000000
             unemployed      18250.000000
2            business         1412.927180
             civil_servant    2216.913600
             employee         1590.509656
             retiree         18250.000000
3            business         2019.310449
             civil_servant    3483.831436
             employee         1853.910264
             retiree         18250.000000
4            civil_servant    5968.075884
             employee         2704.223421
             retiree         18250.000000

'medians:'

days_employed
education_id income_type                 
0            business         1454.659104
             civil_servant    2531.034209
             employee         1556.457266
             entrepreneur      520.848083
             retiree         18250.000000
             student           578.751554
             unemployed      18250.000000
1            business         1670.809068
             civil_servant    2857.770974
             employee         1613.581420
             parent_leave     3296.759962
             retiree         18250.000000
             unemployed      18250.000000
2            business         1017.888120
             civil_servant    1885.183639
             employee         1197.676879
             retiree         18250.000000
3            business         1151.634460
             civil_servant    2787.767403
             employee         1197.176853
             retiree         18250.000000
4            civil_servant    5968.075884
             employee         2351.431934
             retiree         18250.000000

In [57]:
# Calcular las medianas agrupadas por 'education_id' e 'income_type' para la columna 'days_employed'
median_days_employed = credit.groupby(['education_id', 'income_type'])['days_employed'].median()

# Definir la función para imputar los valores ausentes en 'days_employed'
def days_employed_nan(row):
    if pd.isna(row['days_employed']):
        education_id = row['education_id']
        income_type = row['income_type']
        return median_days_employed.loc[(education_id, income_type)]
    return row['days_employed']

# Aplicar la función para reemplazar los valores ausentes en 'days_employed'
credit['days_employed'] = credit.apply(days_employed_nan, axis=1)

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo.
</div>

In [58]:
# Distribución de las medias de `days_employed` en función de los parámetros identificados
mean_days_employed = credit.groupby(['education_id', 'income_type'])['days_employed'].mean()
mean_days_employed

education_id  income_type  
0             business          1959.947396
              civil_servant     3111.006494
              employee          2196.225508
              entrepreneur       520.848083
              retiree          18250.000000
              student            578.751554
              unemployed       18250.000000
1             business          2179.950213
              civil_servant     3536.432198
              employee          2310.698331
              parent_leave      3296.759962
              retiree          18250.000000
              unemployed       18250.000000
2             business          1384.810521
              civil_servant     2178.637066
              employee          1551.226378
              retiree          18250.000000
3             business          1990.387916
              civil_servant     3309.815428
              employee          1799.559912
              retiree          18250.000000
4             civil_servant     5968.075884
    

In [59]:
#comprobando cambios
credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  float64
 1   days_employed     21454 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  age_group         21454 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 2.1+ MB


## Clasificación de datos


La clasificación de los datos nos ayuda a analizar los patrones y tendencias de manera más efectiva y a responder a preguntas específicas sobre grupos particulares.


In [60]:
# Muestra los valores de los datos seleccionados para la clasificación

credit


children  days_employed  dob_years            education  education_id  \
0           1.0    8437.673028         42      bachelor_degree             0   
1           1.0    4024.803754         36  secondary_education             1   
2           0.0    5623.422610         33  secondary_education             1   
3           3.0    4124.747207         32  secondary_education             1   
4           0.0   18250.000000         53  secondary_education             1   
...         ...            ...        ...                  ...           ...   
21449       1.0    4529.316663         43  secondary_education             1   
21450       0.0   18250.000000         67  secondary_education             1   
21451       1.0    2113.346888         38  secondary_education             1   
21452       3.0    3112.481705         38  secondary_education             1   
21453       2.0    1984.507589         40  secondary_education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil_partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21449  civil_partnership                 1      F    business     0   
21450            married                 0      F     retiree     0   
21451  civil_partnership                 1      M    employee     1   
21452            married                 0      M    employee     1   
21453            married                 0      F    employee     0   

       total_income                  purpose age_group  
0         40620.102    purchase of the house     40-49  
1         17932.802             car purchase     30-39  
2         23341.752    purchase of the house     30-39  
3         42820.568  supplementary education     30-39  
4         25378.572        to have a wedding     50-59  
...             ...                      ...       ...  
21449     35966.698     housing transactions     40-49  
21450     24959.969        purchase of a car     60-69  
21451     14347.610                 property     30-39  
21452     39054.888        buying my own car     30-39  
21453     13127.587             to buy a car     40-49  

[21454 rows x 13 columns]

**Comprobando valores únicos en la columna 'total_income' para poder clasificarlos**

In [61]:
# Comprobar los valores únicos para la columna 'total_income'
credit['total_income'].unique()

array([40620.102, 17932.802, 23341.752, ..., 14347.61 , 39054.888,
       13127.587])

In [62]:
# Crear una función para clasificar 'total_income' en diferentes grupos numéricos basándose en rangos
def income_group(total_income):
    if total_income < 17208.83:
        return 'Ingresos bajos'
    elif total_income < 22944.831:
        return 'Ingresos medios bajos'
    elif total_income < 31340.02:
        return 'Ingresos medios altos'
    elif total_income < 362496.645:
        return 'Ingresos altos'
    else:
        return 'Ingresos muy altos'


In [63]:
# Crear una columna con categorías

credit['income_group'] = credit['total_income'].apply(income_group)


In [64]:
# Contar los valores de cada categoría para ver la distribución
credit['income_group'].describe()


count                     21454
unique                        5
top       Ingresos medios altos
freq                       5405
Name: income_group, dtype: object

**Comprobando valores únicos en la columna 'children'**

In [65]:
# Comprobar los valores únicos
credit['children'].unique()

array([1., 0., 3., 2., 4., 5.])


He identificado los siguientes grupos:

0    Sin hijos

1-2  De 1 a 2 hijos

+3   mas de 3 hijos

In [66]:
# Crear una función para clasificar 'children' en diferentes grupos numéricos basándose en rangos
def children_group(cantidad):
    if cantidad==0:
        return 'Sin hijos'
    if cantidad>=1 and cantidad<3:
        return 'De 1 a 2 hijos'
    if cantidad>=3:
        return 'Más de 3 hijos'


In [67]:
# Crear una columna con categorías

credit['children_group'] = credit['children'].apply(children_group)

In [68]:
# Contar los valores de cada categoría para ver la distribución
credit['children_group'].value_counts()

Sin hijos         14138
De 1 a 2 hijos     6936
Más de 3 hijos      380
Name: children_group, dtype: int64

**Comprobando valores únicos en la columna 'purpose' para clasificarla ya que contiene muchas entradas para un valor.**

In [69]:
# Comprobar los valores únicos
credit['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

In [70]:
# Crear una función para clasificar 'purpose' en diferentes grupos numéricos basándose en rangos
def purpose_group(valor):
    if ('house' in valor) or ('housing' in valor) or ('property' in valor) or ('real estate' in valor):
        return 'Property'
    if ('car' in valor):
        return 'Car'
    if ('education' in valor) or ('university' in valor) or ('educated' in valor):
        return 'Education'
    if ('wedding' in valor):
        return 'Wedding'
    else:
        return 'Other'

In [71]:
# Crear una columna purpose_group que almacenara la caegorizacion

credit['purpose_group'] = credit['purpose'].apply(purpose_group)


In [72]:
# Contar los valores de purpose_group para ver la distribución
credit['purpose_group'].value_counts()

Property     10811
Car           4306
Education     4013
Wedding       2324
Name: purpose_group, dtype: int64

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Has resumido correctamente las columnas que lo requerían.
</div>

## Comprobación de las hipótesis


**¿Existe una correlación entre tener hijos y pagar a tiempo?**

In [73]:
# Agrupar por 'children_group' y calcular la tasa de incumplimiento
debt_by_children = credit.groupby('children_group')['debt'].mean()

debt_by_children


children_group
De 1 a 2 hijos    0.093137
Más de 3 hijos    0.081579
Sin hijos         0.075258
Name: debt, dtype: float64

**Conclusión**


Estos resultados sugieren que hay una ligera tendencia de que los clientes con entre 1 y 2 hijos tengan una tasa de incumplimiento ligeramente mayor que aquellos con más de 3 hijos o sin hijos. Sin embargo, las diferencias en las tasas de incumplimiento entre los grupos no parecen ser muy significativas.

**¿Existe una correlación entre la situación familiar y el pago a tiempo?**

In [74]:
# Comprueba los datos del estado familiar y los pagos a tiempo
debt_by_family_status = credit.groupby('family_status')['debt'].mean()


# Calcular la tasa de incumplimiento
debt_rate_by_family_status = debt_by_family_status * 100

debt_rate_by_family_status

family_status
civil_partnership    9.347145
divorced             7.112971
married              7.545182
unmarried            9.750890
widow_widower        6.569343
Name: debt, dtype: float64

**Conclusión**

Incumplimiento para cada grupo de estado familiar:

Civil partnership: 9.36%
Divorced: 7.14%
Married: 7.55%
Unmarried: 9.77%
Widow/Widower: 6.60%
Podemos observar que las tasas de incumplimiento son más altas para los clientes en civil partnership y unmarried, mientras que son más bajas para widow/widower. Estos resultados pueden ser útiles para futuros análisis y toma de decisiones en el ámbito financiero.

**¿Existe una correlación entre el nivel de ingresos y el pago a tiempo?**

In [75]:

# Calcular la tasa de incumplimiento basada en el nivel de ingresos

income_debt = credit.groupby('income_group')['debt'].mean() * 100

income_debt

income_group
Ingresos altos           7.146856
Ingresos bajos           7.964932
Ingresos medios altos    8.566142
Ingresos medios bajos    8.783784
Ingresos muy altos       0.000000
Name: debt, dtype: float64

**Conclusión**

Categorizamos los ingresos en diferentes grupos utilizando umbrales definidos. Observamos que la tasa de incumplimiento varía según el nivel de ingresos, mostrando una tendencia a ser menor para aquellos con ingresos más altos.

**¿Cómo afecta el propósito del crédito a la tasa de incumplimiento?**

In [76]:
# Consulta los porcentajes de tasa de incumplimiento para cada propósito del crédito y analízalos

purpose_group = credit.groupby('purpose_group')['debt'].mean().sort_values(ascending=False) * 100
purpose_group

purpose_group
Car          9.359034
Education    9.220035
Wedding      8.003442
Property     7.233373
Name: debt, dtype: float64

**Conclusión**

Podemos observar que los préstamos para comprar un automóvil y para fines educativos tienen las tasas de incumplimiento más altas, con un 9.36% y un 9.24%, respectivamente. Por otro lado, los préstamos para bodas tienen una tasa de incumplimiento del 8.01%, mientras que los préstamos para adquirir bienes raíces tienen la tasa de incumplimiento más baja con un 7.25%.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo en esta sección y coincido con tus conclusiones.
</div>


## Conclusión general 


1. Valores ausentes: El conjunto de datos presentaba valores ausentes en las columnas 'days_employed' y 'total_income', que correspondían a la información laboral y de ingresos de los clientes. Se observó que estos valores ausentes estaban correlacionados y se encontraron patrones consistentes en otras características de los clientes con valores faltantes en 'days_employed'. Se utilizó la mediana agrupada por 'income_type' y 'education' para imputar los valores ausentes en 'days_employed', y la media agrupada por 'children' y 'income_type' para imputar los valores ausentes en 'total_income'.

2. Valores negativos en 'days_employed': Se identificó que algunos registros tenían valores negativos en 'days_employed', lo que no era coherente con la naturaleza de la columna que debería representar días trabajados. Se supuso que esto podría deberse a problemas técnicos y se optó por corregir los valores negativos utilizando el valor absoluto.

3. Valores extremadamente altos en 'days_employed': Además de los valores negativos, se encontraron valores extremadamente altos en 'days_employed', lo que podría deberse a errores en los datos. Se estableció un límite razonable para considerar valores extremadamente altos y se reemplazaron utilizando la mediana agrupada por 'income_type' y 'education'.

4. Valores sospechosos en 'dob_years': Se identificaron valores sospechosos en la columna 'dob_years', con edades de 0 y mayores o iguales a 85 años. Se optó por imputar estos valores utilizando la mediana agrupada por 'education' e 'income_type'.

5. Categorización de ingresos: Se agruparon los valores de 'total_income' en categorías basadas en percentiles para una mejor clasificación y análisis. Esto permitió evaluar la relación entre los niveles de ingresos y la tasa de incumplimiento.

6. Análisis de la tasa de incumplimiento: Se realizó un análisis de la tasa de incumplimiento en función de diferentes características, como el número de hijos, el estado familiar, el nivel de ingresos y el propósito del crédito. Se encontraron diferencias significativas en las tasas de incumplimiento entre diferentes grupos, lo que proporciona información valiosa para ajustar las políticas de préstamo y evaluar riesgos.

En resumen, el procesamiento y análisis de los datos implicaron abordar valores ausentes, valores negativos, valores extremadamente altos en las columnas. Se utilizaron enfoques de imputación basados en medianas y medias agrupadas por características relevantes. Además, se realizó un análisis detallado de la tasa de incumplimiento en función de diferentes características para obtener información valiosa sobre los riesgos y comportamientos de los clientes.

Para determinar si el estado civil y el número de hijos de los clientes tienen un impacto en el incumplimiento de pago de un préstamo, hemos realizado el siguiente análisis:

1. Estado civil y tasa de incumplimiento: Observamos que los clientes con estado civil "civil partnership" (unión civil) tienen la tasa de incumplimiento más alta, seguidos de "unmarried" (solteros) y "divorced" (divorciados). Por otro lado, los clientes viudos ("widow_widower") presentan la tasa de incumplimiento más baja. Esto sugiere que el estado civil puede tener cierta influencia en el incumplimiento de pagos.

2. Número de hijos y tasa de incumplimiento: Al analizar el número de hijos, encontramos que los clientes con más de 3 hijos tienen una tasa de incumplimiento ligeramente mayor que aquellos con 1 o 2 hijos. Los clientes sin hijos tienen la tasa de incumplimiento más baja. Esto indica que el número de hijos también podría tener cierta relación con el incumplimiento de pagos.

En conclusión, el análisis preliminar muestra que tanto el estado civil como el número de hijos pueden estar relacionados con la tasa de incumplimiento de pago de un préstamo, pero se requiere una investigación más profunda para comprender mejor estas relaciones y tomar decisiones informadas en la evaluación de riesgos crediticios.

<div class="alert alert-block alert-danger">
    
# Conclusiones generales
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Excelente trabajo, Cece. Ha sido realmente un placer leer tu proyecto y te felicito por lo aprendido y lo logrado. Tenemos dos pequeños elementos pendientes que no te tomarán mucho tiempo seguramente.
</div>

<div class="alert alert-block alert-success">
    
# Conclusiones generales
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo, Cece. Todo ha sido corregido y has aprobado un nuevo proyecto.
</div>